src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java

```diff
@@ -1,64 +1,199 @@
 import org.prebid.server.bidder.ttx.proto.TtxImpExt;
 import org.prebid.server.bidder.ttx.proto.TtxImpExtTtx;
+import org.prebid.server.bidder.ttx.response.TtxBidExt;
+import org.prebid.server.bidder.ttx.response.TtxBidExtTtx;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
 import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
 import org.prebid.server.proto.openrtb.ext.request.ttx.ExtImpTtx;
 import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.HttpUtil;
 
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
+import java.util.Objects;
 import java.util.stream.Collectors;
 
-public class TtxBidder extends OpenrtbBidder<ExtImpTtx> {
+public class TtxBidder implements Bidder<BidRequest> {
+
+    private static final TypeReference<ExtPrebid<?, ExtImpTtx>> TTX_EXT_TYPE_REFERENCE =
+            new TypeReference<ExtPrebid<?, ExtImpTtx>>() {
+            };
+
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
 
     public TtxBidder(String endpointUrl, JacksonMapper mapper) {
-        super(endpointUrl, RequestCreationStrategy.SINGLE_REQUEST, ExtImpTtx.class, mapper);
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
     }
 
     @Override
-    protected void modifyRequest(BidRequest bidRequest, BidRequest.BidRequestBuilder requestBuilder,
-                                 List<ImpWithExt<ExtImpTtx>> impsWithExts) {
-        final List<Imp> modifiedImps = impsWithExts.stream()
-                .map(ImpWithExt::getImp)
-                .collect(Collectors.toList());
-        final Imp firstImp = modifiedImps.get(0);
-        final ExtImpTtx firstImpExt = impsWithExts.get(0).getImpExt();
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final HttpRequest<BidRequest> httpRequest = createRequest(request, errors);
 
-        final String zoneId = firstImpExt.getZoneId();
-        final TtxImpExt ttxImpExt = TtxImpExt.of(
-                TtxImpExtTtx.of(firstImpExt.getProductId(), StringUtils.isNotBlank(zoneId) ? zoneId : null));
-
-        final Imp modifiedFirstImp = firstImp.toBuilder().ext(mapper.mapper().valueToTree(ttxImpExt)).build();
-
-        if (modifiedImps.size() == 1) {
-            requestBuilder.imp(Collections.singletonList(modifiedFirstImp));
-        } else {
-            final List<Imp> subList = modifiedImps.subList(1, modifiedImps.size());
-            final List<Imp> finalizedImps = new ArrayList<>(subList.size() + 1);
-            finalizedImps.add(modifiedFirstImp);
-            finalizedImps.addAll(subList);
-            requestBuilder.imp(finalizedImps);
+        if (httpRequest == null) {
```

Comment:
It is better to check errors instead of request.

Reply:
Request can be updated, and errors also present